x-pack/plugins/ml/server/saved_objects/initialization.ts

```diff
@@ -0,0 +1,108 @@
+ */
+
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { IScopedClusterClient, CoreStart } from 'kibana/server';
+import { getInternalSavedObjectsClient } from './util';
+import { repairFactory } from './repair';
+import { jobSavedObjectServiceFactory } from './service';
+import { mlLog } from '../lib/log';
+import { ML_SAVED_OBJECT_TYPE } from '../../common/types/saved_objects';
+
+/**
+ * Creates initializeJobs function which is used to check whether
+ * ml job saved objects exist and creates them if needed
+ *
+ * @param core: CoreStart
+ */
+export function jobSavedObjectsInitializationFactory(core: CoreStart) {
+  const client = (core.elasticsearch.client as unknown) as IScopedClusterClient;
+
+  /**
+   * Check whether ML saved objects exist.
+   * If they don't, check to see whether ML jobs exist.
+   * If jobs exist, but the saved objects do not, create the saved objects.
+   *
+   */
+  async function initializeJobs() {
+    try {
+      if ((await _needsInitializing()) === false) {
+        // ml job saved objects have already been initialized
+        return;
+      }
+      mlLog.info('Initializing job saved objects');
+      const savedObjectsClient = getInternalSavedObjectsClient(core);
+      const jobSavedObjectService = jobSavedObjectServiceFactory(savedObjectsClient);
+      const { initSavedObjects } = repairFactory(client, jobSavedObjectService);
+      await initSavedObjects();
+      mlLog.info('Job saved objects initialized for * space');
+    } catch (error) {
+      mlLog.error('Error Initializing jobs');
+    }
+  }
+
+  async function _needsInitializing() {
+    if (await _jobSavedObjectsExist()) {
+      // at least one ml saved object exists
+      // this has been initialized before
+      return false;
+    }
+
+    if (await _jobsExist()) {
+      // some ml jobs exist, we need to create those saved objects
+      return true;
+    }
+
+    // no ml jobs actually exist,
+    // that's why there were no saved objects
+    return false;
+  }
+
+  async function _jobSavedObjectsExist(size: number = 1) {
+    const { body } = await client.asInternalUser.search({
+      index: '.kibana*',
+      size,
+      body: {
+        query: {
+          bool: {
+            filter: [
+              {
+                term: {
+                  type: ML_SAVED_OBJECT_TYPE,
+                },
+              },
+            ],
+          },
+        },
+      },
+    });
+
+    return body.hits.total.value > 0;
+  }
+
+  async function _jobsExist() {
+    // it would be better to use a simple count search here
+    // but the kibana user does not have access to .ml-config
+    //
+    // const { body } = await client.asInternalUser.count({
+    //   index: '.ml-config',
+    // });
+    // return body.count > 0;
```

Comment:
From the original PR:

@droberts195 @legrego
This function needs to check that there is at least one ML job in existence.
I believe the fastest way to do this would be to perform a count on .ml-config. However, the kibana system user does not have read access to that index.
Instead i'm having to call both AD and DFA get endpoints to perform the same check (below).
Do you think it would be safe to give the kibana system user the ability to make this check?
 
 @legrego legrego 34 minutes ago  Member
I'm not terribly familiar with the contents of .ml-config. It appears to be a listing of jobs? Is there anything else in that index?

A more direct question might be: Is there any information in that index that's not accessible via the ML APIs granted to users with the manage_ml cluster privilege? I'd be ok granting read access to the .ml-config index if this is true, since kibana_system already has that cluster privilege.

On the other hand, if there is information in this index that the kibana_system user currently doesn't have access to via other APIs, then I'd need a better understanding of that data before answering one way or another.

Reply:
@droberts195 what are your thoughts on this?

Reply:
As this PR is holding up a lot of other work, i'm going to leave question open for a future PR.
I'll leave in this commented code in case the decision is to use count.